### PR TITLE
[camerax] Wrap classes to implement resolution configuration for image capture, image analysis, and preview

### PR DESCRIPTION
**Pull request by camsim99**
Wraps classes to implement resolution configuration for image capture, image analysis, and preview. Also bumps CameraX version to latest and removes the deprecated classes used previously.

No functionality changes. Also thanks to @bparrishMines who did majority of the work here!

Part of https://github.com/flutter/flutter/issues/120462

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart [hunk 1]
```diff
@@ -22,6 +22,22 @@ enum CameraStateType {
...
+///    `_LiveDataHostApiImpl#getValueFromInstances` to get the current value of
+///    type S from a LiveData<S> instance and in `LiveDataFlutterApiImpl#create`
+///    to create the expected type of LiveData<S> when requested.
+///
+/// On the native side, ensure the following is done:
+///
+///  * Update `LiveDataHostApiImpl#getValue` is updated to properly return
+///    identifiers for instances of type S.
+///  * Update `ObserverFlutterApiWrapper#onChanged` to properly handle receiving
+///    calls with instances of type S if a LiveData<S> instance is observed.
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart:40]: This is unrelated. Pigeon must have carried over my comment I left in `camerax_library.dart` in a previous PR.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    /** Creates an instance of {@link AspectRatioStrategy}. */
+    @NonNull
+    public AspectRatioStrategy create(
+        @NonNull Long preferredAspectRatio, @NonNull Long fallbackRule) {
+      return new AspectRatioStrategy(preferredAspectRatio.intValue(), fallbackRule.intValue());
+    }
+  }
+
+  /**
+   * Constructs a {@link AspectRatioStrategyHostApiImpl}.
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java:34]: Nit: Constructs an AspectRatioStrategyHostApiImpl

File: packages/camera/camera_android_camerax/test/aspect_ratio_strategy_test.dart [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  TestWidgetsFlutterBinding.ensureInitialized();
+
+  group('AspectRatioStrategy', () {
+    tearDown(() {
+      TestAspectRatioStrategyHostApi.setup(null);
+      TestInstanceManagerHostApi.setup(null);
+    });
+
+    test(
+        'HostApi create makes call to create expected AspectRatioStrategy instance',
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/test/aspect_ratio_strategy_test.dart:28]: Should we have a corresponding detached create test, like we do in the tests for `camera_selector` for example?

File: packages/camera/camera_android_camerax/test/resolution_selector_test.dart [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+void main() {
+  TestWidgetsFlutterBinding.ensureInitialized();
+
+  group('ResolutionSelector', () {
+    tearDown(() {
+      TestResolutionSelectorHostApi.setup(null);
+      TestInstanceManagerHostApi.setup(null);
+    });
+
+    test('HostApi create creates expected ResolutionSelector instance', () {
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/test/resolution_selector_test.dart:31]: Same question, should this also have a corresponding `detached` create test with `verifyNever`

File: packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  /// See https://developer.android.com/reference/androidx/camera/core/resolutionselector/ResolutionStrategy#FALLBACK_RULE_CLOSEST_LOWER_THEN_HIGHER().
+  static const int fallbackRuleClosestLowerThenHigher = 3;
+
+  /// When the specified bound size is unavailable, CameraX falls back to the
+  /// closest lower resolution size.
+  ///
+  /// See https://developer.android.com/reference/androidx/camera/core/resolutionselector/ResolutionStrategy#FALLBACK_RULE_CLOSEST_LOWER()
+  static const int fallbackRuleClosestLower = 4;
+
+  /// Returns the resolution strategey that chooses the highest available
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart:97]: Typo: strategey -> strategy

File: packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import 'java_object.dart';
+
+/// The resolution strategy defines the resolution selection sequence to select
+/// the best size.
+///
+/// See https://developer.android.com/reference/androidx/camera/core/resolutionselector/ResolutionStrategy.
+@immutable
+class ResolutionStrategy extends JavaObject {
+  /// Construct a [ResolutionStrategy].
+  ResolutionStrategy({
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart:19]: What do you think about making the arguments to this constructor non-null (as they are for the android class) and instead using a private named constructor instead for the creation of the highest available strategy?

**Comment by gmackall** [comment]
Left some comments but mostly looks good to me! Main things I'm wondering about are the comment about the constructor for resolution strategy and the question about how we are handling named constants. 

Other than that it's just comment nits and requests for `detached` constructor tests

File: packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart (hunk 6 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart:19]: I love this idea. Changed it and added tests.

File: packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+          instanceManager: instanceManager,
+          binaryMessenger: binaryMessenger,
+        ),
+        super.detached();
+
+  /// CameraX doesn't fall back to select sizes of any other aspect ratio when
+  /// this fallback rule is used.
+  ///
+  /// See https://developer.android.com/reference/androidx/camera/core/resolutionselector/AspectRatioStrategy#FALLBACK_RULE_NONE().
+  static const int fallbackRuleNone = 0;
```
[the remark this answers is not in the document]
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart:76]: My initial thought is that changing them to getters wouldn't be a significant improvement. If the values change, we likely will be quick to find out. Plus, there's the risk that they change the name and a getter would fail there. Not sure which case is more likely but I think it's something to consider. @bparrishMines any thoughts on this?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+   * Constructs an {@link AspectRatioStrategyHostApiImpl}.
+   *
+   * @param instanceManager maintains instances stored to communicate with attached Dart objects
+   */
+  public AspectRatioStrategyHostApiImpl(@NonNull InstanceManager instanceManager) {
+    this(instanceManager, new AspectRatioStrategyProxy());
+  }
+
+  /**
+   * Constructs a {@link AspectRatioStrategyHostApiImpl}.
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java:43]: Nit: Constructs an

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  @VisibleForTesting
+  AspectRatioStrategyHostApiImpl(
+      @NonNull InstanceManager instanceManager, @NonNull AspectRatioStrategyProxy proxy) {
+    this.instanceManager = instanceManager;
+    this.proxy = proxy;
+  }
+
+  /**
+   * Creates a {@link AspectRatioStrategy} instance with the preferred aspect ratio and fallback
```
Comment (gmackall) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/AspectRatioStrategyHostApiImpl.java:56]: Nit: Creates an

File: packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart (hunk 6 above)
Comment (gmackall) [reply packages/camera/camera_android_camerax/lib/src/resolution_strategy.dart:19]: I believe as is, it is still nullable (it is `required` but someone can still pass null). 

I think this can be changed to `required Size this.boundSize`, though, to keep it as an initializing formal but just with a stricter (non-nullable) type. 

See the "NOTE: Also note that it is possible to enforce a type that is stricter"... in https://dart.dev/tools/linter-rules/prefer_initializing_formals (can't link sections unfortunately).

**Review by gmackall — COMMENTED** [review]
Just one more comment about the ResolutionStrategy constructor, and a couple of comment nits!

File: packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (bparrishMines) [reply packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart:76]: It's hard to change them to getters since it would require an async call. Which would make it unintuitive to use. You could also add a Java unit test that verifies that these constants stay the same. However, this would still be susceptible to an instance where the constant values are different for different Android versions. Although this case is probably extremely rare.

**Review by bparrishMines — APPROVED** [review]
LGTM

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 10]
```diff
@@ -116,10 +116,6 @@ class AndroidCameraCameraX extends CameraPlatform {
   @visibleForTesting
   CameraSelector? cameraSelector;
 
-  /// The resolution preset used to create a camera that should be used for
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:119]: Is this a breaking change?

File: packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+  /// CameraX automatically chooses the next best aspect ratio which contains
+  /// the closest field of view (FOV) of the camera sensor, from the remaining
+  /// options.
+  ///
+  /// See https://developer.android.com/reference/androidx/camera/core/resolutionselector/AspectRatioStrategy#FALLBACK_RULE_AUTO().
+  static const int fallbackRuleAuto = 1;
```
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart:83]: Consider bumping these constants to the top of the class for easier discoverability.

File: packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart (hunk 1 above)
Comment (reidbaker) [reply packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart:40]: Should this be removed?

**Review by reidbaker — APPROVED** [review]

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart (hunk 10 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:119]: No, it wasn't used before

File: packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart (hunk 1 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart:40]: No, pigeon will just keep adding it back since it's part of the source file

**Review by gmackall — APPROVED** [review]

File: packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart (hunk 11 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/aspect_ratio_strategy.dart:83]: This would go against the pattern I've been following but if you have strong opinions I can follow this up with another PR moving all static constants to the top of their respective classes
